### PR TITLE
Refactoring abs, all, any and bin

### PR DESCRIPTION
**Pull request by martica**
This updates abs to support any object with an __abs__ method. (And adds an bool.__abs__)
This updates any and all to support sequences defined with __len__ and __getitem__.
This updates bin to support any object with an __index__ method. (And adds bool.__index__ and int.__index__)

Additionally abs, any, all and bin have their __doc__ strings updated from 3.4 to the 3.5-3.7 state.

## PR Checklist:

- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: tests/builtins/test_abs.py [hunk 1]
```diff
@@ -2,6 +2,21 @@
...
+        self.assertCodeExecution("""
+            print(abs.__doc__)
+        """)
+
+    def test_abs_implemented(self):
+        self.assertCodeExecution("""
+            class AbsLike:
+                def __abs__(self):
+                    return 4
+
```
Comment (brutusthebee) [inline tests/builtins/test_abs.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_any.py [hunk 2]
```diff
@@ -24,6 +24,30 @@ def test_any_typeerror(self):
...
+    def test_any_sequence(self):
+        self.assertCodeExecution("""
+            class Sequence:
+              def __init__(self, value):
+                self.value = value
+              def __len__(self):
+                return len(self.value)
+              def __getitem__(self, idx):
+                return self.value[idx]
+
```
Comment (brutusthebee) [inline tests/builtins/test_any.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_any.py [hunk 3]
```diff
@@ -24,6 +24,30 @@ def test_any_typeerror(self):
...
+            print(any(not_all_values))
+            all_values = Sequence([1,2,3,1])
+            print(any(all_values))
+            no_values = Sequence([0,0,0])
+            print(any(no_values))
+        """)
+
+
 
 class BuiltinAnyFunctionTests(BuiltinFunctionTestCase, TranspileTestCase):
```
Comment (brutusthebee) [inline tests/builtins/test_any.py]: At column 1: (E303) too many blank lines (3)

File: tests/builtins/test_all.py [hunk 4]
```diff
@@ -23,6 +23,29 @@ def test_all_typeerror(self):
...
+    def test_all_sequence(self):
+        self.assertCodeExecution("""
+            class Sequence:
+              def __init__(self, value):
+                self.value = value
+              def __len__(self):
+                return len(self.value)
+              def __getitem__(self, idx):
+                return self.value[idx]
+
```
Comment (brutusthebee) [inline tests/builtins/test_all.py]: At column 1: (W293) blank line contains whitespace

File: tests/builtins/test_all.py [hunk 5]
```diff
@@ -23,6 +23,29 @@ def test_all_typeerror(self):
                 print("Done.")
         """)
 
+    def test_all_doc(self):
+        self.assertCodeExecution("""
+            print(all.__doc__)
+        """)
```
Comment (freakboy3742) [inline tests/builtins/test_all.py]: Not necessary for this PR - but couldn't this test be an automated piece of the Builtins test class?

Comment (martica) [reply tests/builtins/test_all.py]: Good call. I'll put it on my list. Probably a lot of ones to update to the new versions.

**Pull request merged**